### PR TITLE
[10.0] base_geoengine fix and improvements

### PR DESCRIPTION
**Pull request by benwillig**
* Implemented convert_to_cache and convert_to_record methods. (Store BaseGeometry hex code in cache and convert it back to BaseGeometry instance when reading from record)
* Added test in convert_to_read method to prevent errors when assigning a BaseGeometry instance to a Geo field.
* New method to create a GeoLine from two points
* Create Geometry column when column is computed but stored
* Fixed tests
* Corrected some pylint issues
* Allow to define which layers will be active on startup
* Allow to show a configurable popup when selecting an item.  Geonengine templates works like kanban box templates.
* Added a hook to allow to set a different style when selecting/hovering a layer
* Allow to change the circle radius with inheritance

### CODE REVIEW

File: base_geoengine/fields.py
Comment (lmignon) [inline base_geoengine/fields.py]: @benwillig ?`else`?

File: base_geoengine/fields.py
Comment (lmignon) [inline base_geoengine/fields.py]: @benwillig Why do you hard code the `srid` value to use in the call to `ST_GeomFromText`?

**Review by lmignon — CHANGES_REQUESTED** [review]

File: base_geoengine/fields.py
Comment (benwillig) [reply base_geoengine/fields.py]: I changed my code. I also discovered a new case which happen during onchange methods. When a onchange computed a GeoPoint, given value in convert_to_record method corresponds to the value returned by the convert_to_read method.

File: base_geoengine/fields.py
Comment (benwillig) [reply base_geoengine/fields.py]: Srid was hardcoded in ```GeoPoint.from_latlon``` method, so I thought there was a reason for that. Should I change the hardcoded ```srid``` in this method too?

File: base_geoengine/__manifest__.py
Comment (dreispt) [inline base_geoengine/__manifest__.py]: Deprecated, remove.

File: base_geoengine/__manifest__.py
Comment (dreispt) [inline base_geoengine/__manifest__.py]: Insn't this deprecated in favour of a extending the assets with QWeb XML?

File: base_geoengine/__manifest__.py
Comment (dreispt) [inline base_geoengine/__manifest__.py]: This should really be the Github repo address, or a link to the docs.

File: base_geoengine/geo_model.py
Comment (dreispt) [inline base_geoengine/geo_model.py]: I suggest using `TODO` in place on `XXX`

File: base_geoengine/security/data.xml
Comment (dreispt) [inline base_geoengine/security/data.xml]: s/admin/Admin

File: base_geoengine/__manifest__.py
Comment (benwillig) [reply base_geoengine/__manifest__.py]: Not for qweb I think, unless I missed something. You extend the asset template only to add javascript and css file.

File: base_geoengine/geo_model.py
Comment (benwillig) [reply base_geoengine/geo_model.py]: I deleted this statement because I already fixed it.

File: base_geoengine/security/data.xml
Comment (benwillig) [reply base_geoengine/security/data.xml]: I don't understand what you mean.

Comment (JonathanNEMRY) [reply base_geoengine/security/data.xml]: @benwillig  I think the idea is to capitalize de "a" of "admin" (s/old_pattern/new_pattern) :smile:

Comment (benwillig) [reply base_geoengine/security/data.xml]: Thanks!

File: base_geoengine/__manifest__.py
Comment (lasley) [reply base_geoengine/__manifest__.py]: * Not deprecated for static QWeb. Even if extending a static QWeb, you would need to include in here except in some special cases where the JS loads it

Comment (dreispt) [reply base_geoengine/__manifest__.py]: ok

File: base_geoengine/fields.py
Comment (lmignon) [inline base_geoengine/fields.py]: @benwillig `srid` can differ by instance. IMO it would be better to add `srid=None` as parameter and if not specified, takes the default one on the class.

**Review by lmignon — CHANGES_REQUESTED** [review]

**Comment by lmignon** [comment]
@benwillig Thank you for all these improvements. I need to  functionally test all this to fully understand the scope of all these changes but these look promising when reading the code. (You can maybe update the documentation with an example of custom popup..) To ease the review and if you plan to add new functionalities, do it in new PR's based on this one.

**Comment by benwillig** [comment]
@yvaucher ping

**Comment by woodbrettm** [comment]
Hi there, any updates on this PR?

File: base_geoengine/static/src/js/views/geoengine_template_widgets.js
Comment (lmignon) [inline base_geoengine/static/src/js/views/geoengine_template_widgets.js]: @benwillig If all this code is new, you could remove N. Bessi as contributor and the Copyright for Camptocamp SA

**Review by lmignon — APPROVED** [review]
LGTM (code review)

File: base_geoengine/__manifest__.py
Comment (yvaucher) [inline base_geoengine/__manifest__.py]: This deserve a version bump

File: base_geoengine/static/src/js/views/geoengine_template_widgets.js
Comment (yvaucher) [reply base_geoengine/static/src/js/views/geoengine_template_widgets.js]: Indeed 
Edit: was replying to @lmignon saying you have to credit only yourself here.

**Review by yvaucher — APPROVED** [review]
Nice work @benwillig 

Thanks for the cleaning and the improvements

"Geonengine templates works like kanban box templates." :heart:

**Comment by yvaucher** [comment]
@benwillig can you just bump the version of the module and change credits so we can merge ?

File: base_geoengine/__manifest__.py
Comment (lasley) [reply base_geoengine/__manifest__.py]: IMO a major version, so 10.0.2.0.0.

**Comment by benwillig** [comment]
Thanks all for your reviews! I updated the version number of the module.

**Comment by benwillig** [comment]
Should I fix the pylint error which cames from ```base_geoengine_demo```? Then I think this PR can be merged if there is no other questions/errors.

**Comment by yvaucher** [comment]
@benwillig 
Here Is see that `base_geoengine` tests don't even start
https://travis-ci.org/OCA/geospatial/jobs/314630058

Can you fix the following error ?

```
MissingError: ('Error, can not automatically initialize spatial postgis support. Database user may have to be superuser and postgres/postgis extentions with their devel header have to be installed. If you do not want Odoo to connect with a super user you can manually prepare your database. To do thisopen a client to your database using a super user and run: \nCREATE EXTENSION postgis;\nCREATE EXTENSION postgis_topology;\n', None)
```

This error seems new in this PR.

For `base_geoengine_demo` no need to fix them here.

**Comment by benwillig** [comment]
Unfortunately, I'm not sure if I can fix it. To create an extension in postgres, the connected user must be a superuser. The only way to fix it is to change the runbot instance creation. I don't know how it was working in previous versions.

**Comment by yvaucher** [comment]
@benwillig it might be a step missing in the .travis.yml

Not sure either why those SQL requests aren't executed with a user which is not superuser. Could be due to version of travis/postgres/MQT

MQT doesn't seems to use a specific DB user:
https://github.com/OCA/maintainer-quality-tools/blob/master/travis/test_server.py#L233

We might be able to fix it by altering `travis` pg user to be superuser.

Could be worth just to do a commit --amend on last commit to trigger a new travis build.

**Comment by benwillig** [comment]
@yvaucher I forced a rebuild but it didn't help. Indeed, if the travis user is used for the DB connection, it should be a postgres superuser.

**Comment by benwillig** [comment]
Any news @yvaucher ?

**Comment by yvaucher** [comment]
@benwillig reading https://docs.travis-ci.com/user/database-setup/#PostgreSQL
 
in .travis.yml try to add:

```
before_script:
  - psql -U postgres -c "create extension postgis"
```

**Comment by benwillig** [comment]
Thanks for your feedback @yvaucher. It works now.

**Comment by yvaucher** [comment]
I would have expected this to work.

Could you try to follow the recommendation here ?

Seems we have the same issue.

https://gis.stackexchange.com/questions/252609/could-not-access-file-libdir-postgis-2-3-on-travis-installation

**Comment by benwillig** [comment]
I think you checked the first job. After it I found the link you gave and I added the postgis package to install in yml package category. The following job suceeded https://travis-ci.org/OCA/geospatial/jobs/367672576

**Comment by yvaucher** [comment]
@benwillig you are right I didn't update this page, it was out of sync.

Thank you for having fixed this, I will then merge it!

**Pull request merged**
